File: ios/MotorModule/MotorKitModule.m
//
//  MotorKitModule.m
//  sdkreactnative
//
//  Created by Prad Nukala on 7/18/22.
//

#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_MODULE(MotorKitModule, RCTEventEmitter)
RCT_EXTERN_METHOD(createAccount:
                  (NSString)password
                  resolve:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject);

RCT_EXTERN_METHOD(checkBridge:
                  (NSString)data
                  resolve:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject);

RCT_EXTERN_METHOD(loginAccount:
                  (NSString)did
                  password: (NSString)password
                  dscKey: (NSDictionary)dscKey
                  pskKey: (NSDictionary)pskKey
                  resolve:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject);
@end

File: ios/MotorModule/MotorKitModule.swift
//
//  MotorKitModule.swift
//  sdkreactnative
//
//  Created by Prad Nukala on 7/18/22.
//

import Foundation
import MotorKit
import React

@objc(MotorKitModule)
class MotorKitModule : RCTEventEmitter {
  @objc override static func requiresMainQueueSetup() -> Bool {
    return true
  }
  override func constantsToExport() -> [AnyHashable : Any]! {
    [:]
  }
  override func supportedEvents() -> [String]! {
    return []
  }
  
  @objc func checkBridge(
    _ data : String,
    resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) -> Void {
    resolve(data)
  }

  @objc func createAccount(
    _ password : String,
    resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) -> Void {
    let motor = MotorKit()
    let account = motor.createAccount(password: password)

    resolve(account)
  }

  @objc(loginAccount:password:dscKey:pskey:)
  public func loginAccount(
    _ did : String, 
    password : String?, 
    dscKey: Data?, 
    pskKey: Data?,
    resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock) -> Void {
    let motor = MotorKit()
    let isLoggedIn = motor.loginAccount(did: did, password: password, dscKey: dscKey, pskKey: pskKey)

    resolve(isLoggedIn)
  }
}
